lib/index.js

```diff
@@ -191,472 +156,477 @@ function Config(options) {
-    if (this._configPaths[bundleName]) {
-        this._configPaths[bundleName][configName] = undefined;
-    }
-};
+    /**
+     * Deregisters a configuration file.
+     * @method deleteConfig
+     * @param {string} bundleName Name of the bundle to which this config file belongs.
+     * @param {string} configName Name of the config file.
+     * @param {string} fullPath Full filesystem path to the config file.
+     * @return {undefined} Nothing appreciable is returned.
+     */
+    deleteConfig: function (bundleName, configName, fullPath) {
+        var bundleMap = this._configPaths[bundleName];
+        if(bundleMap) {
+            var path = bundleMap[configName];
+            if(path && this._configYCBs[path]) {
+                delete this._configYCBs[path];
+            }
+            delete bundleMap[configName];
+            delete this._configIdMap[bundleName][configName];
+            if(Object.keys(bundleMap).length === 0) {
+                delete this._configPaths[bundleName];
+            }
+            if(Object.keys(this._configIdMap[bundleName]).length === 0) {
+                delete this._configIdMap[bundleName];
+            }
+        }
+    },
 
+    /**
+     * Generates a cache key based on config and bundle names, a separator, and a context based key.
+     * Distinct separators can be used to distinguish distinct types of keys, e.g., merged and unmerged reads.
+     *
+     * @param {string} bundleName Name of bundle.
+     * @param {string} separator Separator string to join bundle and config names.
+     * @param {string} configName Name of config.
+     * @param {string} contextKey Key based on the context.
+     * @returns {string} cache key
+     * @private
+     */
+    _getCacheKey: function (bundleName, separator, configName, contextKey) {
+        return bundleName + separator + configName + contextKey;
+    },
 
-/**
- * Reads the contents of the named configuration file.
- * This will auto-detect if the configuration file is YCB and read it in a context-sensitive way if so.
- *
- * This can possibly return the configuration object that is stored in a cache, so the caller should
- * copy it if they intend to make a modifications.
- * @method read
- * @async
- * @param {string} bundleName The bundle in which to find the configuration file.
- * @param {string} configName Which configuration to read.
- * @param {object} [context] The runtime context.
- * @param {Function} callback
- *   @param {Error|null} callback.err If an error occurred, then this parameter will
- *   contain the error. If the operation succeeded, then `err` will be null.
- *   @param {Object} callback.config The merged configuration object, based on the
- *   provided context.
- */
-Config.prototype.read = function (bundleName, configName, context, callback) {
-    var self = this;
-    self._getConfigCache(bundleName, configName, context, true, function (err, config) {
```

Comment:
old code tries to get cached config first before calling `self._getYCB`.  Why does the new code reverse the order?

Reply:
The new code uses the YCB instance itself to generate the context portion of the cache key so we need to get it first. The benefit is the YCB instance has more knowledge about the dimensions used so it can map distinct contexts to same key when they will result in identical read results; we can also avoid the call to sort and it serializes faster.
There's no cost to the order change since the old code looks up the correct cache with the same `bundle -> config -> path -> object` pattern that is used to find the correct YCB instance.